src/main/scala/pbdirect/PBScalarValueReader.scala

```diff
@@ -46,62 +47,123 @@ trait PBScalarValueReaderImplicits_1 {
 
 trait PBScalarValueReaderImplicits extends PBScalarValueReaderImplicits_1 {
 
-  implicit object BooleanReader$ extends PBScalarValueReader[Boolean] {
+  implicit object FunctorReader extends Functor[PBScalarValueReader] {
+    override def map[A, B](reader: PBScalarValueReader[A])(f: A => B): PBScalarValueReader[B] =
+      new PBScalarValueReader[B] {
+        val defaultValue: B                  = f(reader.defaultValue)
+        def canBePacked: Boolean             = reader.canBePacked
+        def read(input: CodedInputStream): B = f(reader.read(input))
+      }
+  }
+
+  implicit val booleanReader: PBScalarValueReader[Boolean] = new PBScalarValueReader[Boolean] {
     override def defaultValue: Boolean                  = false
     override def canBePacked: Boolean                   = true
     override def read(input: CodedInputStream): Boolean = input.readBool()
   }
-  // Stored as variants, but larger in memory: https://groups.google.com/forum/#!topic/protobuf/Er39mNGnRWU
-  implicit object ByteReader$ extends PBScalarValueReader[Byte] {
-    override def defaultValue: Byte                  = 0.toByte
-    override def canBePacked: Boolean                = true
-    override def read(input: CodedInputStream): Byte = input.readInt32().toByte
-  }
-  // Stored as variants, but larger in memory: https://groups.google.com/forum/#!topic/protobuf/Er39mNGnRWU
-  implicit object ShortReader$ extends PBScalarValueReader[Short] {
-    override def defaultValue: Short                  = 0.toShort
-    override def canBePacked: Boolean                 = true
-    override def read(input: CodedInputStream): Short = input.readInt32().toShort
-  }
-  implicit object IntReader$ extends PBScalarValueReader[Int] {
+
+  implicit val untaggedIntReader: PBScalarValueReader[Int] = new PBScalarValueReader[Int] {
     override def defaultValue: Int                  = 0
     override def canBePacked: Boolean               = true
     override def read(input: CodedInputStream): Int = input.readInt32()
   }
-  implicit object LongReader$ extends PBScalarValueReader[Long] {
+
+  implicit val unsignedIntReader: PBScalarValueReader[Int @@ Unsigned] =
+    new PBScalarValueReader[Int @@ Unsigned] {
+      override def defaultValue: Int @@ Unsigned = tag[Unsigned](0)
+      override def canBePacked: Boolean          = true
+      override def read(input: CodedInputStream): Int @@ Unsigned =
+        tag[Unsigned](input.readUInt32())
+    }
```

Comment:
Wondering if we could have some object or abstract class that would receive the types and the function and will create the `PBScalarValueReader` for us

```scala
def apply[A, B](read: CodedInputstream => A): PBScalarValueReader[A @@ B] = ???
```